crates/bevy_render/src/batching/mod.rs

```diff
@@ -1,87 +1,46 @@
 
 /// Add this component to mesh entities to disable automatic batching
 #[derive(Component)]
 pub struct NoAutomaticBatching;
 
-pub trait BatchMeta<T: BatchMeta<T>> {
-    fn matches(&self, other: &T) -> bool;
-}
-
-struct BatchState<T: BatchMeta<T>> {
-    meta: Option<T>,
-    /// The base index in the object data binding's array
-    index: Option<NonMaxU32>,
-    /// The dynamic offset of the data binding
-    dynamic_offset: Option<NonMaxU32>,
-    /// The number of entities in the batch
-    count: u32,
-    item_index: usize,
-}
-
-impl<T: BatchMeta<T>> Default for BatchState<T> {
-    fn default() -> Self {
-        Self {
-            meta: Default::default(),
-            index: Default::default(),
-            dynamic_offset: Default::default(),
-            count: Default::default(),
-            item_index: Default::default(),
-        }
-    }
-}
-
-fn update_batch_data<I: PhaseItem, T: BatchMeta<T>>(item: &mut I, batch: &BatchState<T>) {
-    let BatchState {
-        count,
-        index,
-        dynamic_offset,
-        ..
-    } = batch;
-    let index = index.map_or(0, |index| index.get());
-    *item.batch_range_mut() = index..(index + *count);
-    *item.dynamic_offset_mut() = *dynamic_offset;
-}
-
 /// Batch the items in a render phase. This means comparing metadata needed to draw each phase item
 /// and trying to combine the draws into a batch.
 pub fn batch_render_phase<
     I: CachedRenderPipelinePhaseItem,
-    T: BatchMeta<T>, // Batch metadata used for distinguishing batches
+    T: PartialEq, // Batch metadata used for distinguishing batches
 >(
     phase: &mut RenderPhase<I>,
-    mut get_batch_meta: impl FnMut(&I) -> Option<(T, Option<NonMaxU32>, Option<NonMaxU32>)>,
+    mut get_batch_meta: impl FnMut(&I) -> Option<(T, NonMaxU32, Option<NonMaxU32>)>,
 ) {
-    let mut batch = BatchState::<T>::default();
-    for i in 0..phase.items.len() {
-        let item = &phase.items[i];
-        let Some((batch_meta, index, dynamic_offset)) = get_batch_meta(item) else {
-            // It is necessary to start a new batch if an entity not matching the query is
-            // encountered. This can be achieved by resetting the batch meta.
-            batch.meta = None;
+    // we iterate in reverse so that we can write to the last item of the current batch, and still skip the 
+    // right number of phase items when iterating forwards in the Render stage
+    let mut items = phase.items.iter_mut().rev().peekable();
+    let mut batch_start_index = None;
+    let mut next_batch = items.peek().and_then(|item| get_batch_meta(item));
+    while let Some(item) = items.next() {
+        // get current batch meta and update next batch meta
+        let Some((batch_meta, index, dynamic_offset)) = std::mem::replace(
+            &mut next_batch,
+            items.peek().and_then(|item| get_batch_meta(item)),
+        ) else {
+            // if the current phase item doesn't match the query, we don't modify it
```

Comment:
It isn't clear to me if/how this code handles the case of an entity in the phase not matching the query used in the get_batch_meta closure splitting this batch. For transparent phases it is important that things are drawn in back to front order according to the sorted phase. If we batch things returned by get_batch_meta() across the items where it returns None then we would violate that. Sorry if I'm just not seeing it and the code already takes care of this. :)

Reply:
so if we have say [match1, match2, no match, match3], the code will go

loop 1: (batch_meta, ..) = match1, next_batch = match2
-> set batch start index
-> match1 matches match2 => skip update

loop 2: (batch_meta, ..) = match2, next_batch = None
-> (don't set batch start index as it's already set)
-> match2 doesn't match None => update with batch start index and current index == 0..=1

loop 3: (batch_meta, ..) fails, next_batch = match 3
-> continue

loop 4: (batch_meta, ..) = match3, next_batch = None
-> set batch start index
-> match3 doesn't match None => update with batch start index and current index == 2..=2

so i think it should be ok?

Reply:
and the phase.items look like:

data  = [match3, no match, match2, match1]


initially:   [phase item 1 = match3, phase item 2 = no match, phase item 3 = match2, phase item 4 = match1]
after loop1: [phase item 1 = match3, phase item 2 = no match, phase item 3 = match2, phase item 4 *unchanged*]
after loop2: [phase item 1 = match3, phase item 2 = no match, phase item 3 *= 0..=1*, phase item 4 unchanged]
after loop3: [phase item 1 = match3, phase item 2 *unchanged*, phase item 3 = 0..=1, phase item 4 unchanged]
after loop4: [phase item 1 *=2..=2*, phase item 2 unchanged, phase item 3 = 0..=1, phase item 4 unchanged]

so you're right there is an issue - because i iterate backwards the indexes don't line up between the GpuArrayBuffer indexes and the phase.items indexes. in this case phase item 2 has GpuArrayIndex 2 but will have item index 1... oops

Reply:
sorry no i take it back, if the item doesn't match then the initial batch range will be 0..1 which would still be correct. if the get_batch_meta returns None then the mesh doesn't get pushed on to the GpuArrayBuffer.

so i think it's all fine... but if the test didn't work then obviously not?

Reply:
sorry for the flurry of comments.

> For transparent phases it is important that things are drawn in back to front order according to the sorted phase

i think this is a problem though. it'll end up running batches in the right/sorted order, but reversing the items within the batch.

Reply:
ok i pushed a change to remove the reverse. now i really reckon it'll work. but please run your interleaved test (or point me at it)